File: lab04/Trapecio.java
import java.util.*;
import java.math.BigDecimal;
import java.math.RoundingMode;

interface Funcion {
    double evaluar(double x);
}

final class FuncionCuadratica implements Funcion {
    public double evaluar(double x) { return 2.0*x*x + 3.0*x + 0.5; }
}

class Tarea implements Runnable {
    private final Funcion f;
    private final double a;
    private final double h;
    private final long inicio, fin;
    private double resultado;

    Tarea(Funcion f, double a, double h, long inicio, long fin) {
        this.f = f; this.a = a; this.h = h; this.inicio = inicio; this.fin = fin;
    }
    public void run() {
        double suma = 0.0;
        for (long i = inicio; i <= fin; i++) {
            double x = a + i * h;
            suma += f.evaluar(x);
        }
        resultado = suma;
    }
    public double get() { return resultado; }
}

public class Trapecio {
    private static String clave(double v) {
        return new BigDecimal(v).setScale(12, RoundingMode.HALF_UP).toPlainString();
    }
    private static double integrar(Funcion f, double a, double b, long n, int hilos) throws Exception {
        double h = (b - a) / n;
        long interior = Math.max(0, n - 1);
        long base = interior / hilos;
        long resto = interior % hilos;
        List<Thread> ts = new ArrayList<>(hilos);
        List<Tarea> tareas = new ArrayList<>(hilos);
        long inicio = 1;
        for (int w = 0; w < hilos; w++) {
            long tam = base + (w < resto ? 1 : 0);
            long fin = (tam == 0) ? (inicio - 1) : (inicio + tam - 1);
            Tarea t = new Tarea(f, a, h, inicio, fin);
            tareas.add(t);
            Thread th = new Thread(t);
            ts.add(th);
            th.start();
            inicio = fin + 1;
        }
        for (Thread th : ts) th.join();
        double suma = 0.0;
        for (Tarea t : tareas) suma += t.get();
        double fa = f.evaluar(a), fb = f.evaluar(b);
        return h * (0.5 * fa + suma + 0.5 * fb);
    }
    public static void main(String[] args) throws Exception {
        double a = (args.length >= 1) ? Double.parseDouble(args[0]) : 2.0;
        double b = (args.length >= 2) ? Double.parseDouble(args[1]) : 20.0;
        long Nmax = (args.length >= 3) ? Long.parseLong(args[2]) : 2_000_000L;
        int hilos = (args.length >= 4) ? Integer.parseInt(args[3]) : Math.max(1, Runtime.getRuntime().availableProcessors());
        Funcion f = new FuncionCuadratica();
        String ultimaClave = null;
        double ultimoValor = Double.NaN;
        long repetido = -1;
        long n;
        long t0 = System.nanoTime();
        for (n = 1; n <= Nmax; n++) {
            double val = integrar(f, a, b, n, hilos);
            String k = clave(val);
            if (k.equals(ultimaClave)) { repetido = n; ultimoValor = val; break; }
            ultimaClave = k;
            ultimoValor = val;
            if (n <= 10 || n % 100000 == 0) {
                System.out.printf(Locale.US, "n=%d -> %.12f%n", n, val);
            }
        }
        long t1 = System.nanoTime();
        if (repetido > 0) System.out.printf(Locale.US, "Repetido en n=%d -> %.12f%n", repetido, ultimoValor);
        else System.out.printf(Locale.US, "Finalizado en n=%d -> %.12f%n", n - 1, ultimoValor);
        System.out.printf(Locale.US, "Tiempo: %.3f s%n", (t1 - t0)/1e9);
    }
}

File: lab04/trapecio.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <iomanip>
#include <sstream>
#include <chrono>
#include <limits>
#include <algorithm>

using namespace std;

struct Funcion {
    virtual double eval(double x) const = 0;
    virtual ~Funcion() = default;
};

struct FuncionCuadratica : Funcion {
    double eval(double x) const override {
        return 2.0 * x * x + 3.0 * x + 0.5;
    }
};

struct Tarea {
    const Funcion* f;
    double a;
    double h;
    long long inicio;
    long long fin;
    double resultado;

    void operator()() {
        double suma = 0.0;
        for (long long i = inicio; i <= fin; ++i) {
            double x = a + i * h;
            suma += f->eval(x);
        }
        resultado = suma;
    }
};

static string clave(double v) {
    ostringstream oss;
    oss.setf(std::ios::fixed);
    oss << setprecision(12) << v;
    return oss.str();
}

double integrar(const Funcion& f, double a, double b, long long n, size_t hilos) {
    double h = (b - a) / static_cast<double>(n);
    long long interior = max(0LL, n - 1);
    long long base = interior / static_cast<long long>(hilos);
    long long resto = interior % static_cast<long long>(hilos);

    vector<thread> th;
    th.reserve(hilos);
    vector<Tarea> tareas(hilos);

    long long inicio = 1;
    for (size_t w = 0; w < hilos; ++w) {
        long long tam = base + (static_cast<long long>(w) < resto ? 1 : 0);
        long long fin = (tam == 0) ? (inicio - 1) : (inicio + tam - 1);
        tareas[w] = Tarea{&f, a, h, inicio, fin, 0.0};
        th.emplace_back(ref(tareas[w]));
        inicio = fin + 1;
    }

    for (auto& t : th) t.join();

    double suma = 0.0;
    for (auto& t : tareas) suma += t.resultado;

    double fa = f.eval(a);
    double fb = f.eval(b);
    return h * (0.5 * fa + suma + 0.5 * fb);
}

int main(int argc, char** argv) {
    double a = (argc >= 2) ? atof(argv[1]) : 2.0;
    double b = (argc >= 3) ? atof(argv[2]) : 20.0;
    long long Nmax = (argc >= 4) ? atoll(argv[3]) : 2000000LL;
    size_t hilos = (argc >= 5) ? static_cast<size_t>(atoll(argv[4]))
                               : max(1u, thread::hardware_concurrency());

    FuncionCuadratica f;

    cout.setf(std::ios::fixed);
    cout << setprecision(12);

    string ultima;
    double valor = numeric_limits<double>::quiet_NaN();
    long long repetido = -1;

    auto t0 = chrono::steady_clock::now();
    for (long long n = 1; n <= Nmax; ++n) {
        double v = integrar(f, a, b, n, hilos);
        string k = clave(v);
        if (n == 1) {
            ultima = k;
            valor = v;
        } else {
            if (k == ultima) {
                repetido = n;
                valor = v;
                break;
            }
            ultima = k;
            valor = v;
        }
        if (n <= 10 || n % 100000 == 0) {
            cout << "n=" << n << " -> " << v << "\n";
        }
    }
    auto t1 = chrono::steady_clock::now();
    double s = chrono::duration<double>(t1 - t0).count();

    if (repetido > 0) {
        cout << "Repetido en n=" << repetido << " -> " << valor << "\n";
    } else {
        cout << "Finalizado -> " << valor << "\n";
    }
    cout << "Tiempo: " << s << " s\n";
    return 0;
}
